File: leetcode/101_symmetric_tree.py
"""
101. Symmetric Tree
- Easy
- Tree, DFS, BFS
- Link: https://leetcode.com/problems/symmetric-tree/
"""


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        return self.helper(root, root)
    
    def helper(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q: return True
        if not p or not q: return False
        
        return p.val == q.val and self.helper(p.left, q.right) and self.helper(p.right, q.left)

File: leetcode/101_symmetric_tree.java
// """
// 101. Symmetric Tree
// - Easy
// - Tree, DFS, BFS
// - Link: https://leetcode.com/problems/symmetric-tree/
// """

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public boolean isSymmetric(TreeNode root) {
      return helper(root, root);
  }
  
  public boolean helper(TreeNode p, TreeNode q) {
      if (p == null && q == null) return true;
      if (p == null || q == null) return false;
      
      return p.val == q.val && helper(p.left, q.right) && helper(p.right, q.left);
  }
}
